src/Test/Inspection/Core.hs

```diff
@@ -203,43 +204,49 @@ eqSlice it slice1 slice2
 -- | Returns @True@ if the given core expression mentions no type constructor
 -- anywhere that has the given name.
 freeOfType :: Slice -> [Name] -> Maybe (Var, CoreExpr)
-freeOfType slice tcNs = allTyCons (\tc -> getName tc `notElem` tcNs) slice
-
-allTyCons :: (TyCon -> Bool) -> Slice -> Maybe (Var, CoreExpr)
-allTyCons predicate slice = listToMaybe [ (v,e) | (v,e) <- slice, not (go e) ]
+freeOfType slice tcNs =
+      fmap (\(a,b,_) -> (a,b))
+    $ allTyCons (\tc -> getName tc `notElem` tcNs) slice
+
+-- | Check if all type constructors in a slice satisfy the given predicate.
+-- Returns the binder, expression and failing constructors triple on failure.
+allTyCons :: (TyCon -> Bool) -> Slice -> Maybe (Var, CoreExpr, [TyCon])
+allTyCons predicate slice =
+    listToMaybe
+        [(v, e, nub tcs) | (v, e) <- slice, let tcs = go e, not (null tcs)]
```

Comment:
A `Slice` is a list of binders, so it is a bit odd to return just _one_ binder, but _all_ tycons therein: Either report the first error (one binder, one tc), or all binders and all tcs. I think the former should suffice. Would also avoid the (O(n)!) `nub. Something like the simple and elegant
```suggestion
        [(v, e, tc) | (v, e) <- slice, tc <- go e]
```
(with just `TyCon`, not `[TyCon]` in the type for `allTycons`).

Reply:
I thought it may be helpful in practice to report all the tycons because we can report that without using a lot of real estate in the output, I know we are not doing that for binders but it does not cause any harm other than annoying the perfectionist mind. I am neutral on this change. You can change it before merging.